ledger/ledger-on-sql/src/main/scala/com/daml/ledger/on/sql/SqlLedgerReaderWriter.scala

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.on.sql
+
+import java.sql.Connection
+import java.time.Clock
+import java.util.UUID
+
+import akka.NotUsed
+import akka.stream.Materializer
+import akka.stream.scaladsl.Source
+import com.daml.ledger.on.sql.SqlLedgerReaderWriter._
+import com.daml.ledger.on.sql.queries.Queries
+import com.daml.ledger.on.sql.queries.Queries.Index
+import com.daml.ledger.participant.state.kvutils.DamlKvutils.{
+  DamlLogEntryId,
+  DamlStateKey,
+  DamlStateValue,
+  DamlSubmission
+}
+import com.daml.ledger.participant.state.kvutils.api.{LedgerReader, LedgerRecord, LedgerWriter}
+import com.daml.ledger.participant.state.kvutils.{Envelope, KeyValueCommitting}
+import com.daml.ledger.participant.state.v1._
+import com.digitalasset.daml.lf.data.Ref
+import com.digitalasset.daml.lf.data.Time.Timestamp
+import com.digitalasset.daml.lf.engine.Engine
+import com.digitalasset.ledger.api.health.{HealthStatus, Healthy}
+import com.digitalasset.platform.akkastreams.dispatcher.Dispatcher
+import com.digitalasset.platform.akkastreams.dispatcher.SubSource.RangeSource
+import com.google.protobuf.ByteString
+import com.zaxxer.hikari.{HikariConfig, HikariDataSource}
+import javax.sql.DataSource
+
+import scala.collection.JavaConverters._
+import scala.concurrent.{ExecutionContext, Future}
+import scala.util.Random
+
+class SqlLedgerReaderWriter(
+    ledgerId: LedgerId = Ref.LedgerString.assertFromString(UUID.randomUUID.toString),
+    val participantId: ParticipantId,
+    queries: Queries,
+    connectionSource: DataSource with AutoCloseable,
+)(implicit executionContext: ExecutionContext, materializer: Materializer)
+    extends LedgerWriter
+    with LedgerReader
+    with AutoCloseable {
+
+  private val engine = Engine()
+
+  private val dispatcher: Dispatcher[Index] =
+    Dispatcher(
+      "sql-participant-state",
+      zeroIndex = FirstIndex,
+      headAtInitialization = FirstIndex,
+    )
+
+  private val randomNumberGenerator = new Random()
+
+  // TODO: implement
+  override def checkHealth(): HealthStatus = Healthy
+
+  override def close(): Unit = {
+    dispatcher.close()
+    connectionSource.close()
+  }
+
+  override def retrieveLedgerId(): LedgerId = ledgerId
+
+  override def events(offset: Option[Offset]): Source[LedgerRecord, NotUsed] =
+    dispatcher
+      .startingAt(
+        offset.getOrElse(FirstOffset).components.head,
+        RangeSource((start, end) =>
+          withConnection { implicit connection =>
+            Future(Source(queries.selectFromLog(start, end)))
+        })
+      )
+      .map { case (_, record) => record }
+
+  override def commit(correlationId: String, envelope: Array[Byte]): Future[SubmissionResult] =
+    withConnection { implicit connection =>
```

Comment:
The operations below (read / update state) should be performed in a single transaction.

Reply:
They are. We `setAutoCommit(false)` below, which means everything happens in a transaction. Acquiring the connection starts the transaction, and closing it implicitly commits. I'm not a big fan of implicit behavior, so [`withConnection`](https://github.com/digital-asset/daml/pull/3940/files#diff-9f45aadbe621258428069b8880a87344R160) explicitly commits if all goes well. It also catches exceptions and explicitly rolls back.

Reply:
Let's rename this to `inTransaction`.

Reply:
I suggest adding `DB` or `Database` (e.g. `inDbTransaction`) to the name as well. There are enough different kinds of transactions that warrant making it super clear which one is meant.

Reply:
Thanks for the clarification @SamirTalwar.

Reply:
Renamed it to `inDatabaseTransaction`, and renamed the other `withConnection` to `withDatabaseStream`.